### PR TITLE
Add 2 VAPI endpoints

### PR DESCRIPTION
**Pull request by ChrisGuzman**
- [x] Add a call modifier endpoint
- [x] Add the DTMF endpoint

### CODE REVIEW

File: src/main/java/com/nexmo/client/dtmf/SendDTMFMethod.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.util.EntityUtils;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+
+public class SendDTMFMethod extends AbstractMethod<DTMFRequest, DTMFResponse> {
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/dtmf/SendDTMFMethod.java]: @judy2k I copied your pattern of extending `AbstractMethod` from the voice client methods. Is there another way I should do this?

File: src/main/java/com/nexmo/client/dtmf/DTMFPayload.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+public class DTMFPayload {
+    private String digits;
```
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/dtmf/DTMFPayload.java]: The [DTMF docs](https://docs.nexmo.com/voice/voice-api/api-reference#dtmfparameters) says that it takes "The array of digits to send to the Call" but looking at the implementations for ruby and python, it looks like they're sending a string. Do you know which is correct?

File: src/main/java/com/nexmo/client/dtmf/SendDTMFMethod.java (hunk 1 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/dtmf/SendDTMFMethod.java]: I think that's correct in this case. You'd only want to implement your own Method from scratch if you wanted to bypass the authentication mechanism, or do something else weird ;-)

File: src/main/java/com/nexmo/client/dtmf/DTMFPayload.java (hunk 2 above)
Comment (judy2k) [inline src/main/java/com/nexmo/client/dtmf/DTMFPayload.java]: I'd follow the existing implementations, and maybe file a documentation ticket. The implementations have the advantage of being tested :wink:

**Comment by ChrisGuzman** [comment]
@judy2k I've noticed that in the `Call.java` class the `to` and `answerUrl` are Strings not Lists are the API page suggests https://docs.nexmo.com/voice/voice-api/api-reference#ccparameters. I haven't tested this branch against the API but wanted to bring it to your attention. Maybe we need to change the attributes in the `Call.java` class?

**Comment by leggetter** [comment]
Is the workflow and signatures of other APIs/resources/sub-resources similar to this? e.g.

1. Create a request object
2. Upon creation pass a payload object with the optional request parameters?
3. Once you've made the request you get a response object

**Comment by leggetter** [comment]
Would it be possible to update the README to show what the VAPI functionality is going to look like? I appreciate that may change, but I'd just like to get an idea of what we're likely aiming for.

**Comment by ChrisGuzman** [comment]
@leggetter asked:
> Is the workflow and signatures of other APIs/resources/sub-resources similar to this? e.g.
> Create a request object
> Upon creation pass a payload object with the optional request parameters?
> Once you've made the request you get a response object

In my experience, yes most SDKs for APIs have a similar workflow.

**Comment by judy2k** [comment]
It's a limitation of the language. Java doesn't have default arguments for params, or dictionary literals, so depending on the complexity of the call, if you have any optional arguments you either provide multiple overloaded methods, or you provide a configurable request object.

The whole thing can be made slightly nicer by adding Builders for the request objects, but I haven't done that (yet - I'm probably going to).

**Comment by ChrisGuzman** [comment]
@judy2k I'm at -2.71% code coverage. Can you help me with the last bit of tests I should write or should we merge it and worry about adding additional tests later.

**Comment by ChrisGuzman** [comment]
Actually before we merge it I guess it would be good to do a proper code review. I think some of the models (eg: `Payload.java`) and some of the methods in the tests could be better named.

File: src/main/java/com/nexmo/client/dtmf/DTMFEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package com.nexmo.client.dtmf;/*
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/dtmf/DTMFEndpoint.java]: I think everything related to dtmf should be under the voice package. I don't think dtmf code makes sense outside the context of VAPI? (Specifically, I think this should be `com.nexmo.client.voice.dtmf`)

File: src/main/java/com/nexmo/client/voice/DTMFPayload.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package com.nexmo.client.voice;/*
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/DTMFPayload.java]: The formatting of these copyright comments starting on the same line as the package declaration is a mistake in my IntelliJ template. Feel free to put them before the package declaration if you prefer - I think this is what we agreed the other day.

File: src/main/java/com/nexmo/client/voice/DTMFRequest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.nexmo.client.NexmoUnexpectedException;
+
+public class DTMFRequest {
+    private String uuid;
+    private DTMFPayload payload;
+
+    public DTMFRequest(String uuid, DTMFPayload payload) {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/DTMFRequest.java]: Would it be worth adding a convenient constructor with the following signature?

```java
public DTMFRequest(String uuid, String digits)
```

I figure that would be nicer for the user, given that the DTMFPayload only really exists as an implementation detail, for JSON serialization - in fact, maybe remove this constructor entirely and have the DTMFPayload object initialized inside the new constructor?

File: src/main/java/com/nexmo/client/voice/NexmoVoiceClient.java [hunk 6]
```diff
@@ -48,4 +51,13 @@ public CallRecordPage listCalls() throws IOException, NexmoClientException {
     public CallRecordPage listCalls(CallsFilter filter) throws IOException, NexmoClientException {
         return calls.get(filter);
     }
+
+    public void sendDTMF(DTMFRequest dtmfRequest) throws IOException, NexmoClientException {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/NexmoVoiceClient.java]: I think maybe either replace or overload this method with the signature below? What do you think?

```java
sendDTMF(String uuid, String digits)
```

File: src/main/java/com/nexmo/client/voice/Payload.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+public class Payload {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/Payload.java]: The name `Payload` is very generic - every endpoint that accepts `PUT` or `POST` requests takes a payload. Maybe `ModifyCallPayload`?

File: src/main/java/com/nexmo/client/voice/CallModifier.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+public class CallModifier {
+    private String uuid;
+    private Payload payload;
+
+    public CallModifier(String uuid, Payload payload) {
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/CallModifier.java]: I feel the same way here about `Payload` as I do about `DTMFPayload` - it's an internal detail, and ideally we should hide it from the user. `CallModifier(String uuid, String action)` instead?

I was wondering if maybe we should have a package for internal-only objects. What do you think is the best way to "hide" them from the user?

File: src/main/java/com/nexmo/client/dtmf/DTMFEndpoint.java (hunk 3 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/dtmf/DTMFEndpoint.java]: Agreed

File: src/main/java/com/nexmo/client/voice/DTMFPayload.java (hunk 4 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/DTMFPayload.java]: Will change

File: src/main/java/com/nexmo/client/voice/Payload.java (hunk 7 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/Payload.java]: `ModifyCallPayload` sounds good

File: src/main/java/com/nexmo/client/voice/DTMFRequest.java (hunk 5 above)
Comment (ChrisGuzman) [inline src/main/java/com/nexmo/client/voice/DTMFRequest.java]: Yes I think you're right

**Comment by ChrisGuzman** [comment]
@judy2k Can you review changes again and approve if everything's fine?

File: src/main/java/com/nexmo/client/voice/CallRecord.java [hunk 9]
```diff
@@ -178,7 +178,7 @@ public static CallRecord fromJson(String json) {
             mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
             return mapper.readValue(json, CallRecord.class);
         } catch (IOException jpe) {
-            throw new NexmoUnexpectedException("Failed to produce json from Call object.", jpe);
+            throw new NexmoUnexpectedException("Failed to produce json from CallRecord object.", jpe);
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/CallRecord.java]: Nice catch.

File: src/main/java/com/nexmo/client/voice/SendDTMFMethod.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        put.setHeader("Content-Type", "application/json");
+        put.setEntity(new StringEntity(request.toJson()));
+        LOG.info("Request: " + request.toJson());
+        return put;
+    }
+
+    @Override
+    public DTMFResponse parseResponse(HttpResponse response) throws IOException {
+        String json = EntityUtils.toString(response.getEntity());
+        LOG.info("Received: " + json);
```
Comment (judy2k) [inline src/main/java/com/nexmo/client/voice/SendDTMFMethod.java]: I've been changing my log messages like this to DEBUG, but it's a task of its own, so feel free to leave this now and we can tidy up in a later PR.

File: src/test/java/com/nexmo/client/voice/CallModifierTest.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    @Test
+    public void setAction() throws Exception {
+        //this is not a real action
+        callModifier.setAction("pause");
+        assertEquals("pause", callModifier.getAction());
+    }
+
+    @Test
+    public void toJson() throws Exception {
+        String jsonString = "{" + "\"action\":\"hangup\"" + "}";
```
Comment (judy2k) [inline src/test/java/com/nexmo/client/voice/CallModifierTest.java]: What's the reason for defining this as three string literals?

File: src/test/java/com/nexmo/client/voice/endpoints/ModifyCallMethodTest.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
+package com.nexmo.client.voice.endpoints;
+
+import com.auth0.jwt.internal.com.fasterxml.jackson.databind.JsonNode;
```
Comment (judy2k) [inline src/test/java/com/nexmo/client/voice/endpoints/ModifyCallMethodTest.java]: Be really careful about this - we have two different Jacksons available - one is in `com.fasterxml.jackson.databind.JsonNode`, and the other vendored into `com.auth0.jwt.internal.com.fasterxml.jackson.databind.JsonNode`. (My IDE kept auto-completing to the wrong one)

File: src/test/java/com/nexmo/client/voice/CallModifierTest.java (hunk 11 above)
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/voice/CallModifierTest.java]: I think it was a bad copy/paste. I'll clean it up

File: src/test/java/com/nexmo/client/voice/endpoints/ModifyCallMethodTest.java (hunk 12 above)
Comment (ChrisGuzman) [inline src/test/java/com/nexmo/client/voice/endpoints/ModifyCallMethodTest.java]: Doh

**Pull request merged**